File: editors/vscode/src/constructs/imports.ts
// The TypeScript half of the `memql/imports` contract (memql#3335).
//
// THE LANGUAGE SERVER OWNS ALL .memql PARSING. This module exists to make that
// true of the one place it was not: run/bundle.ts walked the import graph with
// a line regex, because there was no import request to ask. There is now.
//
// A regex cannot know what the lexer knows. A `use` line inside a /* ... */
// block comment is not an import, and the scan followed it anyway; the failure
// in the other direction -- a dirty dependency missed, so the run executes the
// SAVED copy of a file the developer is editing -- is the invisible class this
// epic kept tripping over.
//
// The wire shape below is a FIXED contract mirrored from cmd/memql-lsp/imports.go.
// Field names and the "empty array, never null" rule on `imports` and `names`
// are load-bearing; changing either means changing both sides in one commit.
//
// Deliberately free of `vscode` imports so it is unit-testable under bare
// `node --test`; extension.ts is the adapter that sends the request. The rule
// is enforced mechanically -- see cmd/memql-lsp/vscodeimportrule_test.go.

// -----------------------------------------------------------------------------
// The wire contract
// -----------------------------------------------------------------------------

export interface MemqlImport {
  /**
   * The dotted MODULE path as written -- "cognition.concepts".
   *
   * It names a FILE (dsl/cognition/concepts.memql), not a construct.
   * Resolving it onto disk stays on this side: the workspace layout being
   * walked is the client's.
   */
  path: string;
  /**
   * The imported-construct list from the brace group, in authored order.
   *
   * The bundle walk does not use it -- a file with any unsaved edit goes into
   * the bundle whole, whichever of its constructs are referenced -- but it is
   * what the server has, and dropping it here would mean a second round trip
   * for the first consumer that wants it.
   */
  names: string[];
}

// The server capability key. Feature-detect on this rather than calling the
// request blind and handling MethodNotFound -- an older memql-lsp on the PATH
// is an ordinary situation (serverPath is a user setting and the PATH fallback
// is whatever is installed), not an error worth a popup mid-run.
export const IMPORTS_CAPABILITY = "memqlImports";
export const IMPORTS_METHOD = "memql/imports";

// -----------------------------------------------------------------------------
// Defensive decoding
// -----------------------------------------------------------------------------

/**
 * parseImports narrows the raw JSON-RPC result.
 *
 * The language server is a separate process, possibly an OLDER BUILD than this
 * extension, so the reply is untrusted in the ordinary sense. Every
 * malformation degrades to "this is not an import" rather than throwing: a
 * thrown error here would abort a run the developer just started, and the
 * degraded outcome -- a dependency left out of the bundle -- is the same
 * bounded cost the walk has always had for an import it could not resolve.
 */
export function parseImports(raw: unknown): MemqlImport[] {
  if (raw === null || typeof raw !== "object") return [];
  const imports = (raw as { imports?: unknown }).imports;
  if (!Array.isArray(imports)) return [];
  const out: MemqlImport[] = [];
  for (const entry of imports) {
    const parsed = parseOne(entry);
    if (parsed !== undefined) out.push(parsed);
  }
  return out;
}

/**
 * importPaths is the bundle walk's view: the distinct module paths, in first
 * appearance order.
 *
 * Deduplicated HERE rather than on the server, which reports declarations as
 * authored. Two `use` lines naming one module are two declarations and one
 * file, and the walk wants the file.
 */
export function importPaths(imports: readonly MemqlImport[]): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const imp of imports) {
    if (seen.has(imp.path)) continue;
    seen.add(imp.path);
    out.push(imp.path);
  }
  return out;
}

function parseOne(raw: unknown): MemqlImport | undefined {
  if (raw === null || typeof raw !== "object") return undefined;
  const r = raw as Record<string, unknown>;
  const path = r.path;
  if (typeof path !== "string" || path === "") return undefined;
  // The contract promises an array, but a missing or malformed `names` says
  // nothing about whether the module is imported -- and the walk does not read
  // names at all. Degrading to [] keeps the dependency rather than dropping it
  // over a field this consumer does not use.
  const names = Array.isArray(r.names)
    ? r.names.filter((n): n is string => typeof n === "string")
    : [];
  return { path, names };
}

File: component/memql/sense/imports.go
package sense

// imports.go answers one question for an editor client: which files does this
// buffer import? It is the language-server half of `memql/imports`
// (memql#3335), and it exists to delete a SECOND PARSER.
//
// The VS Code runtime panel assembles the bundle it session-defines from the
// active file plus any transitively `use`-imported workspace file that
// currently has unsaved edits. To walk that graph it was scanning for
// `use <dotted>.{` with a line regex -- a regex that cannot know what a lexer
// knows. The design's first principle is that there is exactly one parser, and
// the failure mode of the second one is the invisible kind: a missed dirty
// dependency means the run silently executes the SAVED copy of a file the
// developer is looking at.
//
// Two things the regex gets wrong that this does not:
//
//  1. A `use` line inside a /* ... */ BLOCK COMMENT. The lexer skips block
//     comments outright, so a commented-out import produces no tokens at all;
//     a line-anchored regex sees an ordinary import and follows it.
//  2. Line structure generally. The lexer does not care where the newlines
//     are, so a declaration split across lines -- or two on one line -- reads
//     identically to the compiler and to this.
//
// WHY A TOKEN WALK RATHER THAN A FULL PARSE. Diagnose parses the whole file
// through applyRewriteChain, and one half-typed construct anywhere makes that
// fail. The editor asks this question during a run, on a buffer that is
// usually mid-edit, and "your imports vanished because line 200 is
// incomplete" is exactly the invisible failure the request is meant to remove.
// So this walks the LEXER's token stream: the real lexer, with comments and
// string literals already resolved, but tolerant of a body it never inspects.
// That keeps the single-parser guarantee where it matters -- what counts as an
// import is decided by the compiler's own tokenizer -- while surviving the
// buffer states the client actually asks about.
//
// Only Form B (`use <module>.{ names }`) is recognised, because that is the
// only import form the loader accepts: Form A (`use <ns>.<concept>`, with or
// without `as`) is retired and rejected at parse time, so following one would
// mean following something that cannot load.

import "github.com/znasllc-io/memql/component/language/parser"

// Import is one file-top Form-B `use` declaration, as authored.
//
// The wire projection of this type is a FIXED contract shared with the VS Code
// extension (memql#3335); field semantics must not drift without changing both
// sides together.
type Import struct {
	// Path is the dotted MODULE path as written: "cognition.concepts",
	// "common.traits". It names a FILE -- `cognition.shapes` is
	// dsl/cognition/shapes.memql -- not a construct. Resolving it to a path on
	// disk is the client's job: the workspace layout it is walking is the
	// client's, and a language server answering for a document it was never
	// told about would have to guess at it.
	Path string
	// Names is the imported-construct list from the brace group, in authored
	// order. Carried because it is free and a client may want it (a future
	// "which construct came from where" affordance); the bundle walk itself
	// only needs Path -- a file with any unsaved edit goes into the bundle
	// whole, whichever of its constructs are used.
	Names []string
}

// Imports returns every Form-B `use` declaration in source, in declaration
// order, INCLUDING duplicates of the same module path.
//
// Duplicates are kept because this reports what the buffer says rather than a
// deduplicated view of it: two `use` lines naming one module with different
// brace lists are two declarations, and a client that only wants the distinct
// module set can collapse them in one line. Collapsing here would throw away
// the Names of all but the first.
//
// It never returns an error. A buffer that does not lex yields an empty
// result, for the same reason RunnableConstructs does: the editor asks this on
// a mid-edit buffer as the ordinary case, and an error would surface as a
// popup on a document the developer is simply still typing.
//
// The returned slice is always non-nil.
func (s *Service) Imports(source string) []Import {
	out := []Import{}
	lexer := parser.NewLexer(source)
	tokens, err := lexer.Tokenize()
	if err != nil {
		return out
	}
	for i := 0; i < len(tokens); i++ {
		if tokens[i].Type != parser.TokenKeywordUse {
			continue
		}
		imp, next, ok := importAt(tokens, i)
		if !ok {
			// Not a well-formed Form-B declaration. Skipped rather than
			// guessed at: `use` is a reserved keyword, so this is either the
			// retired Form A (which cannot load, so following it would walk to
			// a file the compiler will not read) or a half-typed line the
			// developer is still writing.
			continue
		}
		out = append(out, imp)
		i = next
	}
	return out
}

// importAt matches `use <dotted> . { name, name }` starting at the `use` token
// at tokens[i]. It returns the projected declaration and the index of the
// closing brace, so the caller resumes after the whole declaration rather than
// re-entering it.
//
// The shape it matches is exactly the one parseUseDeclaration accepts, token
// for token. Worth knowing about the lexer here: a dotted path lexes as ONE
// identifier -- `cognition.concepts` is a single token -- because the
// identifier scanner consumes a dot followed by an identifier character. The
// trailing dot before `{` is deliberately NOT consumed by that scanner, which
// is what makes the `.` `{` pair a reliable separator.
func importAt(tokens []parser.Token, i int) (Import, int, bool) {
	j := i + 1
	if j >= len(tokens) || tokens[j].Type != parser.TokenIdentifier {
		return Import{}, 0, false
	}
	path := tokens[j].Literal
	j++
	if j >= len(tokens) || tokens[j].Type != parser.TokenDot {
		return Import{}, 0, false
	}
	j++
	if j >= len(tokens) || tokens[j].Type != parser.TokenBraceOpen {
		return Import{}, 0, false
	}
	j++

	names := []string{}
	for ; j < len(tokens); j++ {
		switch tokens[j].Type {
		case parser.TokenComma, parser.TokenSemicolon:
			continue
		case parser.TokenIdentifier:
			names = append(names, tokens[j].Literal)
			continue
		case parser.TokenBraceClose:
			// An empty brace list is rejected by the parser
			// ("must list at least one imported name"), so it is not a
			// declaration the loader would accept either.
			if len(names) == 0 {
				return Import{}, 0, false
			}
			return Import{Path: path, Names: names}, j, true
		}
		// Anything else inside the braces means this is not the declaration it
		// looked like -- most often a buffer mid-edit where the brace group is
		// not closed yet.
		return Import{}, 0, false
	}
	return Import{}, 0, false
}
